File: LeetCode/Study Plan/Practice Problems/Medium/RLE Iterator/Cpp/rle_iterator.cc
#include <iostream>
#include <vector>

typedef std::vector<int> vi;

class RLEIterator {
    vi v;
    RLEIterator(vector<int>& encoding) {
        this->v = encoding;
    }
    
    int next(int n) {
        if (this->v.size() > 0) {
			for (int i=0; i<this->v.size(); i+=2) {
				int cnt = this->v[i];
				int val = this->v[i+1];
				if (cnt >= n) {
					this->v[i] -= n;
					return val;
				}else {
					n-= this->v[i];
					this->v[i] = 0;
				}
			}
        }
		return -1;
    }
};

File: LeetCode/Medium/RLE Iterator/Go/rle_iterator.go
package main

type RLEIterator struct {
	arr []int
}

func Constructor(encoding []int) RLEIterator {
	return RLEIterator{arr: encoding}
}

func (this *RLEIterator) Next(n int) int {
	if len(this.arr) > 0 {
		for i := 0; i < len(this.arr); i += 2 {
			cnt := this.arr[i]
			val := this.arr[i+1]
			if cnt >= n {
				this.arr[i] -= n
				return val
			} else {
				n -= this.arr[i]
				this.arr[i] = 0
			}
		}
	}
	return -1
}
